x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/state/state.go

```diff
@@ -0,0 +1,162 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build windows
+
+package state
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/registry"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/tables"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/filters"
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/logger"
+)
+
+// Expiration duration for the global state.
+const defaultExpirationDuration = 3 * time.Minute
+
+// Default path to the Amcache hive.
+const defaultHivePath = "C:\\Windows\\AppCompat\\Programs\\Amcache.hve"
+
+// GlobalState is a shared state object that holds cached Amcache entries.
+// It is needed because loading and parsing the Amcache hive can be slow,
+// so we want to avoid doing it on every query. Instead, we load it once and
+// cache the results for a configurable duration.
+//
+// the amcache hive has multiple tables that can be queried, and when joins
+// are performed, osquery may call the generate function for each table multiple
+// times per query.  This can lead to significant performance issues if the
+// hive is reloaded each time.  By caching the results in a global state object,
+// we can avoid this overhead.
+//
+// The cache is not populated until the first query is made, to avoid
+// unnecessary work if the tables are not used.  Additionally, the cache
+// is refreshed at query time if it has expired. But will not be updated until
+// the next query, even if it is expired.
+
+type cachedTables map[tables.TableName][]tables.Entry
+
+func newCachedTables() cachedTables {
+	cachedTables := cachedTables{}
+	for _, amcacheTable := range tables.AllAmcacheTables() {
+		cachedTables[amcacheTable.Name] = nil
+	}
+	return cachedTables
+}
+
+type AmcacheState struct {
+	cache              cachedTables
+	hivePath           string
+	expirationDuration time.Duration
+	lock               sync.RWMutex
+	timer              *time.Timer
+}
+
+// Global variables for the gInstance and a mutex to protect it.
+var (
+	instance *AmcacheState
+)
+
+// newAmcacheState creates a nnw AmcacheState instance with the default configuration.
+func newAmcacheState(hivePath string, expirationDuration time.Duration) *AmcacheState {
+	state := &AmcacheState{hivePath: hivePath, expirationDuration: expirationDuration, cache: nil}
+	state.timer = time.AfterFunc(expirationDuration, func() {
```

Comment:
Nothing ever cleans up this timer, which is a resource leak (of a goroutine).

```
func (gs *AmcacheState) Close() {
    gs.lock.Lock()
    defer gs.lock.Unlock()
    if gs.timer != nil {
        gs.timer.Stop()
        gs.timer = nil
    }
    gs.cache = nil
}
```

You can call that from tests with a `defer`, and it looks like there is shutdown hook in `main_windows.go` where you can wire this in.